numpyro/primitives.py

```diff
@@ -232,7 +233,20 @@ def process_message(self, msg):
         cond_indep_stack = msg['cond_indep_stack']
         frame = CondIndepStackFrame(self.name, self.dim, self.subsample_size)
         cond_indep_stack.append(frame)
-        batch_shape = self._get_batch_shape(cond_indep_stack)
+        expected_shape = self._get_batch_shape(cond_indep_stack)
+        dist_batch_shape = msg['fn'].batch_shape if msg['type'] == 'sample' else ()
+        overlap_idx = len(expected_shape) - len(dist_batch_shape)
+        if overlap_idx < 0:
+            raise ValueError('Expected dimensions within plate = {}, which is less than the '
+                             'distribution\'s batch shape = {}.'.format(len(expected_shape), len(dist_batch_shape)))
+        trailing_shape = expected_shape[overlap_idx:]
+        # e.g. distribution with batch shape (1, 5) cannot be broadcast to (5, 5)
+        broadcast_shape = lax.broadcast_shapes(trailing_shape, dist_batch_shape)
+        if sum(map(sub, broadcast_shape, dist_batch_shape)) > 0:
```

Comment:
As I understand, broadcast_shape[i] >= dist_batch_shape[i] for each index `i`. So this sum greater than 0 is equivalent to the equality of two shapes. I would prefer to compare `broadcast_shape == dist_batch_shape` directly.

Reply:
Note that broadcasting up in particular cannot be done, but the reverse is allowed:

```python
with plate('plate', 10, dim=-2):
  pyro.sample('b', dist.Bernoulli(np.array([0.5, 0.5]))) # batch shape = (2,) ; trailing plate = (1,)
```

This is what we are doing in the infer test below actually. Directly comparing will also fail one of the tests in `test_handlers`.

Reply:
But `trailing_shape=(1,)` will be broadcasted to `dist_batch_shape=(2,)` in that case, so we have
`broadcast_shape=(2,)`, `dist_batch_shape=(2,)`. Do I understand it correctly?

Reply:
My bad, I got confused with my own broadcasting manipulations! 😄 You are right, we can directly compare and it will fail the other case where we are trying to broadcast up.